### ISSUE TITLE
Using DOM API in modules too

### ISSUE THREAD

**Issue by katyo** [snabbdom/snabbdom#258]
Introducing DOM API is much useful idea but I think that the modules also must modify nodes using it.

I propose to add `domapi` arg to some hooks, which can be used to patch real DOM nodes. Also it need to add appropriate functions to DOM API implementations, like getAttribute, setAttribute, removeAttribute, getProperty, setProperty, removeProperty, addClass, removeClass, ... and so on.

I try to implement server-side rendering which will operates through native snabbdom's `patch` call using special DOM API implementation. Currently I need implement special DOM Node objects with prototypes (i.e. class instances in OOP terms), because currently no way to avoid it.

**Comment by acstll** [comment]
I'm curious, why aren't you using [`snabbdom-to-html`](https://github.com/snabbdom/snabbdom-to-html) package to render server-side?

In the beginning, we tried monkey patching the DOM API, mainly inspired by this: https://www.npmjs.com/package/html-element but it didn't worked out well.

**Comment by mightyiam** [comment]
Maybe look at how Cycle.js does server-side? I’ve easily implemented server-side rendering with it. Here’s the module:

https://github.com/cyclejs/cyclejs/blob/9dddacff477b58a14b8c089ebf96c0f3dbe3bde6/dom/src/makeHTMLDriver.ts

**Comment by katyo** [comment]
@acstll The `snabbdom-to-html` package currently has too many useless dependencies.
@mightyiam Currently I implemented this feature in much like way, but I think this way is not so proper way at all, according to the law of leaky abstractions.

**Comment by acstll** [comment]
@katyo "useless" doesn't seem to me like the right word here, maybe "unnecessary" or "redundant"?

Plus, why is it relevant if the package has 10 dependencies instead of 5, or 8, or 20, if you're using it in node to do server-side rendering and *weight* is not really a problem?

Anyway, I'm obviously open to suggestions and PRs to make it better so it could eventually suite your needs as well.

**Comment by katyo** [comment]
@acstll Of course, I really mean "redundant" here. Dependencies is not so bad when it helps us to solve complex problems in simple ways. But it not so good when we make simple things.
I planned run server code on embedded platform with simple ES5 interpreter, i.e. not nodejs, so the dependencies is important factor for me. The snabbdom is a dependency-less library, so I found is useful in that context.

**Comment by acstll** [comment]
@katyo nice to know, thank you for the explanation :^)

**Comment by caesarsol** [comment]
The idea is legit, could every module have an init which accepts the domapi?
Any other dom-access leak should be addressed then.

**Comment by caridy** [comment]
> The idea is legit, could every module have an init which accepts the domapi?

That will make the module code stateful, which is not desirable.
